File: 42-trapping-rain-water/solution.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        if (height.size() == 0)   {
            return 0;
        }

        int l = 0;
        int r = height.size() - 1;
        int left_max = height[0];
        int right_max = height[height.size()-1];
        int result = 0;

        while(l < r)    {
            if(left_max < right_max)  {
                l += 1;
                left_max = max(left_max, height[l]);
                result += left_max - height[l];
            }
            else    {
                r -= 1;
                right_max = max(right_max, height[r]);
                result += right_max - height[r];
            }
        }
        return result;
        //Time Complexity: O(n)
        //Space Complexity: O(1)
    }
};

File: 42-trapping-rain-water/solution.py
class Solution:
    def trap(self, height: List[int]) -> int:
        waterCount = 0
        for x in range(len(height)):
            if x == 0:
                lmax = 0
            else: 
                lmax = max(height[0:x])
            if x == len(height)-1:
                rmax = 0
            else:
                rmax = max(height[x+1:len(height)])
            tempCount = (min(rmax, lmax) - height[x])
            if tempCount < 0:
                tempCount = 0
            waterCount += tempCount
        return waterCount
            